// @expo/image-utils/build/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Image_1 = require("./Image");
exports.generateFaviconAsync = Image_1.generateFaviconAsync;
exports.generateImageAsync = Image_1.generateImageAsync;
const jimp_1 = require("./jimp");
exports.jimpAsync = jimp_1.jimpAsync;
const sharp_1 = require("./sharp");
exports.findSharpInstanceAsync = sharp_1.findSharpInstanceAsync;
exports.isAvailableAsync = sharp_1.isAvailableAsync;
exports.sharpAsync = sharp_1.sharpAsync;
const Cache = __importStar(require("./Cache"));
exports.Cache = Cache;
function imageAsync(options, commands = []) {
    return __awaiter(this, void 0, void 0, function* () {
        if (yield sharp_1.isAvailableAsync()) {
            return sharp_1.sharpAsync(options, commands);
        }
        return jimp_1.jimpAsync(Object.assign(Object.assign({}, options), { format: jimp_1.convertFormat(options.format), originalInput: options.input }), commands);
    });
}
exports.imageAsync = imageAsync;
